### DISCUSSION TITLE
XOR with Zod in Prisma 4.11 with `Without<...>` type

### DISCUSSION THREAD

**Discussion by chrishoermann** [colinhacks/zod#2171] [Q&A]
I've built a generator for prisma ([zod-prisma-types](https://github.com/chrishoermann/zod-prisma-types)) that throws a type error since prisma updated their `XOR` type in 4.11:

```ts
  type Without<T, U> = { [P in Exclude<keyof T, keyof U>]?: never };

  /**
   * XOR is needed to have a real mutually exclusive union type
   * https://stackoverflow.com/questions/42123407/does-typescript-support-mutually-exclusive-types
   */
  type XOR<T, U> =
    T extends object ?
    U extends object ?
      (Without<T, U> & U) | (Without<U, T> & T)
    : U : T
```

before this type update the following zod schema worked without problems:

```ts
export const UserCreateInputSchema: z.ZodType<Prisma.UserCreateInput> = z.object({
  id: z.string().cuid().optional(),
  email: z.string().email({ message: "Invalid email address" }),
  name: z.string().min(1).max(100).optional().nullable(),
  role: z.union([ z.lazy(() => UserCreateroleInputSchema),z.lazy(() => RoleSchema).array() ]).optional(),
  enum: z.lazy(() => AnotherEnumSchema).optional(),
  scalarList: z.union([ z.lazy(() => UserCreatescalarListInputSchema),z.string().array() ]).optional(),
  posts: z.lazy(() => PostCreateNestedManyWithoutAuthorInputSchema).optional(),
  profile: z.lazy(() => ProfileCreateNestedOneWithoutUserInputSchema).optional(),
  location: z.lazy(() => LocationCreateNestedOneWithoutUserInputSchema).optional(),
}).strict();

export const UserUncheckedCreateInputSchema: z.ZodType<Prisma.UserUncheckedCreateInput> = z.object({
  id: z.string().cuid().optional(),
  email: z.string().email({ message: "Invalid email address" }),
  name: z.string().min(1).max(100).optional().nullable(),
  role: z.union([ z.lazy(() => UserCreateroleInputSchema),z.lazy(() => RoleSchema).array() ]).optional(),
  enum: z.lazy(() => AnotherEnumSchema).optional(),
  scalarList: z.union([ z.lazy(() => UserCreatescalarListInputSchema),z.string().array() ]).optional(),
  lat: z.number(),
  lng: z.number(),
  posts: z.lazy(() => PostUncheckedCreateNestedManyWithoutAuthorInputSchema).optional(),
  profile: z.lazy(() => ProfileUncheckedCreateNestedOneWithoutUserInputSchema).optional(),
}).strict();

export const UserCreateArgsSchema: z.ZodType<Prisma.UserCreateArgs> = z.object({
  select: UserSelectSchema.optional(),
  include: UserIncludeSchema.optional(),
  data: z.union([ UserCreateInputSchema,UserUncheckedCreateInputSchema ]),
}).strict()
```

so a simple union was sufficient to satisfy typescript. in Prisma 4.11 typescript now throws the following error because, as I see it, the `Without<...>` type can not be satisfied with a union anymore:
```
Type 'ZodObject<{ select: ZodOptional<ZodType<UserSelect, ZodTypeDef, UserSelect>>; include: ZodOptional<ZodType<UserInclude, ZodTypeDef, UserInclude>>; data: ZodUnion<...>; }, "strict", ZodTypeAny, { ...; }, { ...; }>' is not assignable to type 'ZodType<UserCreateArgs, ZodTypeDef, UserCreateArgs>'.
  The types of '_type.data' are incompatible between these types.
    Type '(UserCreateInput | UserUncheckedCreateInput) & (UserCreateInput | UserUncheckedCreateInput | undefined)' is not assignable to type '(Without<UserCreateInput, UserUncheckedCreateInput> & UserUncheckedCreateInput) | (Without<...> & UserCreateInput)'.
      Type 'UserCreateInput & UserUncheckedCreateInput' is not assignable to type '(Without<UserCreateInput, UserUncheckedCreateInput> & UserUncheckedCreateInput) | (Without<...> & UserCreateInput)'.
        Type 'UserCreateInput & UserUncheckedCreateInput' is not assignable to type 'Without<UserUncheckedCreateInput, UserCreateInput> & UserCreateInput'.
          Type 'UserCreateInput & UserUncheckedCreateInput' is not assignable to type 'Without<UserUncheckedCreateInput, UserCreateInput>'.
            Types of property 'lat' are incompatible.
              Type 'number' is not assignable to type 'undefined'.
```

I'm currently a bit stuck on how to solve this issue because I don't really know how to implement the `Without<...>` type especially because the `z.ZodType<...>` does not provide the object helper methods that would have been my first guess on how to approach this. Any help is appreciated.

**Comment by JacobWeisenburger** [comment]
Perhaps something like this?

```ts
type Foo = {
    foo: string
    bar: number
}

const schema = z.object( {
    foo: z.string(),
    bar: z.number(),
} ) satisfies z.ZodType<Foo>

schema.pick // exists
schema.omit // exists
schema.extend // exists
schema.merge // exists
schema.partial // exists
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by chrishoermann** [reply]
@JacobWeisenburger thanks for the idea but this sadly would not work since some of the generated schemas need an explicit type annotation due to their recursive nature. But I investigated a bit and it seems that this is actually a prisma/typescript issue because the same `XOR` type is generated in 4.10.1 and there it works. So this type can't be the cause of the issue. I'll put together a testrepo tomorrow and see what the prisma and/or typescript guys have to say about this.

**Reply by njdowdy** [reply]
FWIW, I tested various combinations of packages in my projects. The culprit for me is zod 3.21.2. The changelog for that is [here](https://github.com/colinhacks/zod/releases/tag/v3.21.2).

Everything is fine with this combination:
  typescript: 4.9.5
  prisma: 4.11.0
  zod: 3.21.1
  zod-prisma-types: 2.4.1

I hope that helps troubleshoot.

**Reply by chrishoermann** [reply]
@njdowdy Big thanks, 🙏  I also was playing around with versions but focused my search more on the prisma side. But now that we know that I will create an issue from this discussion.

**Reply by ymc9** [reply]
Thank you both @chrishoermann @njdowdy . You guys saved my day!

**Reply by samthompsonkennedy** [reply]
> FWIW, I tested various combinations of packages in my projects. The culprit for me is zod 3.21.2. The changelog for that is [here](https://github.com/colinhacks/zod/releases/tag/v3.21.2).
> 
> Everything is fine with this combination: typescript: 4.9.5 prisma: 4.11.0 zod: 3.21.1 zod-prisma-types: 2.4.1
> 
> I hope that helps troubleshoot.

Can confirm, that this is what's required to get `zod-prisma-types`, those versions saved the day for 3 projects now!

**Comment by erik-metz** [comment]
With the package combination:
typescript: 4.9.5
prisma: 4.11.0
zod: 3.21.1
zod-prisma-types: 2.4.1

I'm currently getting this error:
`Type 'ZodObject<{ where: ZodLazy<ZodType<UserWhereUniqueInput, ZodTypeDef, UserWhereUniqueInput>>; create: ZodUnion<...>; }, "strict", ZodTypeAny, { ...; }, { ...; }>' is not assignable to type 'ZodType<UserCreateOrConnectWithoutAccountsInput, ZodTypeDef, UserCreateOrConnectWithoutAccountsInput>'.
  The types of '_type.create' are incompatible between these types.
    Type '(UserCreateWithoutAccountsInput | UserUncheckedCreateWithoutAccountsInput) & (UserCreateWithoutAccountsInput | UserUncheckedCreateWithoutAccountsInput | undefined)' is not assignable to type '(Without<UserCreateWithoutAccountsInput, UserUncheckedCreateWithoutAccountsInput> & UserUncheckedCreateWithoutAccountsInput) | (Without<...> & UserCreateWithoutAccountsInput)'.
      Type 'UserCreateWithoutAccountsInput & UserUncheckedCreateWithoutAccountsInput' is not assignable to type '(Without<UserCreateWithoutAccountsInput, UserUncheckedCreateWithoutAccountsInput> & UserUncheckedCreateWithoutAccountsInput) | (Without<...> & UserCreateWithoutAccountsInput)'.
        Type 'UserCreateWithoutAccountsInput & UserUncheckedCreateWithoutAccountsInput' is not assignable to type 'Without<UserUncheckedCreateWithoutAccountsInput, UserCreateWithoutAccountsInput> & UserCreateWithoutAccountsInput'.
          Type 'UserCreateWithoutAccountsInput & UserUncheckedCreateWithoutAccountsInput' is not assignable to type 'Without<UserUncheckedCreateWithoutAccountsInput, UserCreateWithoutAccountsInput>'.
            Types of property 'companyId' are incompatible.
              Type 'string | undefined' is not assignable to type 'undefined'.
                Type 'string' is not assignable to type 'undefined'.`

can somone suuport?

**Comment by artem-alek** [comment]
any updates on this. I am in this same issue...

**Comment by valueof** [comment]
In our case the reason was that `strictNullChecks` was `false` in TS config.
